README.md
# active_storage

Send files to Rails Active Storage using direct upload.

[![CI](https://github.com/algonauti/dart-active-storage/workflows/CI/badge.svg)](https://github.com/algonauti/dart-active-storage/actions)

## Getting Started

First, you instantiate an `ActiveStorage` object, passing it your Rails' backend 
url where `ActiveStorage::DirectUploadsController` (or your customization of it) 
is responding.  

```dart
var activeStorage = ActiveStorage(
  directUploadURL: 'http://localhost:3000/rails/active_storage/direct_uploads',
);
```

Then, you just have to invoke the `upload()` method, passing it a `File` 
instance along with its info, like this:

```dart
DirectUploadResponse response = await activeStorage.upload(
  fileName: 'picture.jpg',
  fileMimeType: 'image/jpeg',
  file: File('my/picture.jpg'),
);
```

The returned `DirectUploadResponse` object provides a `signedId` getter. You 
should set that value in the API request attribute that your Rails backend will
be setting on the `ActiveRecord` model attribute "powered by" `ActiveStorage` 
(i.e. annotated with `has_one_attached`). 

## Custom Headers

It's very likely that your Rails backend's upload endpoint is accessible to 
authenticated users only. In that case, you'll need to pass an `Authorization`
header together with the upload request. To do so, you just have to invoke the
`addHeader()` method on the `ActiveStorage` instance, before invoking the 
`upload()` method. Like this:

```dart
activeStorage.addHeader(
  'Authorization',
  'Bearer ae38e518bfb4e2b11c2e5cefffe4581e66af07296cfec4b2299ab54a74ef3d8c',
);
```

Another case when you might want to set a custom header is for adding 
`User-Agent` information to your upload requests.

## Upload Progress

When uploading large files, it's important to be able to provide some feedback 
to your users about the progress. To do so, you have the option to pass a 
callback function to the `upload()` method, which would get called as many times
as the number of byte chunks that are sent to the server, and every time would 
receive a `double` representing the upload progress percentage. Like this:

```dart
DirectUploadResponse response = await activeStorage.upload(
  fileName: 'large.jpg',
  fileMimeType: 'image/jpeg',
  file: File('pics/large.jpg'),
  onProgress: (percent) {
    print('Uploaded ${percent.toStringAsFixed(2)}%');
  },
);
```


CHANGELOG.md
## [1.1.0] - Jun 4th, 2022

* Improved error handling
* Updated packages

## [1.0.0] - Aug 24th, 2021

* Migrated to null safety

## [0.2.0] - May 6rd, 2021

* Upgraded `http` to `0.13.3`
* Upgraded `crypto` to `3.0.1`

## [0.1.0] - November 26th, 2019

* Initial release


test/utils.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart';

/// The current server instance.
HttpServer? _server;

/// The URL for the current server instance.
Uri get serverUrl => Uri.parse('http://localhost:${_server!.port}');

/// Starts a new HTTP server.
Future startServer() {
  return HttpServer.bind("localhost", 0).then((s) {
    _server = s;
    s.listen((request) {
      var method = request.method;
      var path = request.uri.path;
      var response = request.response;

      if (path == '/error') {
        response.statusCode = 400;
        response.contentLength = 0;
        response.close();
        return;
      }

      if (method == 'POST' && path == '/direct-upload') {
        response.statusCode = 200;
        response.headers.set('Content-Type', 'application/json; charset=utf-8');
        var responseData = {
          'id': 153,
          'key': 'zcYSmqkyKqJH6Mez7WJAcuJ8',
          'filename': 'avalon.jpg',
          'content_type': 'image/jpeg',
          'metadata': {},
          'byte_size': 88037,
          'checksum': 'BdXPe8f6bkr3J4PcnYgnxw==',
          'created_at': '2019-06-12T15:41:29.105Z',
          'signed_id':
              'eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWms9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--e9e436de22c082cd1142a42ea29daa7eaf5df97d',
          'direct_upload': {
            'url': '$serverUrl/file-upload',
            'headers': {'Content-MD5': 'BdXPe8f6bkr3J4PcnYgnxw=='}
          }
        };
        var body = json.encode(responseData);
        response.contentLength = body.length;
        response.write(body);
        response.close();
        return;
      }

      if (method == 'PUT' && path == '/file-upload') {
        ByteStream(request).toBytes().then((requestBodyBytes) {
          if (requestBodyBytes.isEmpty) {
            response.statusCode = 403;
          } else {
            response.statusCode = 200;
          }
          response.close();
        });
      }
    });
  });
}

/// Stops the current HTTP server.
void stopServer() {
  if (_server != null) {
    _server!.close();
    _server = null;
  }
}


test/uploader_test.dart
import 'dart:io';

import 'package:active_storage/active_storage.dart';
import 'package:active_storage/src/uploader.dart';
import 'package:test/test.dart';

import 'utils.dart';

void main() {
  group('directUpload()', () {
    test('returns expected response data', () {
      return startServer().then((_) {
        var uploader = Uploader('$serverUrl/direct-upload');
        return uploader.directUpload(directUploadRequestSample);
      }).then((DirectUploadResponse responseData) {
        expect(responseData.id, 153);
        expect(responseData.key, 'zcYSmqkyKqJH6Mez7WJAcuJ8');
        expect(
          responseData.signedId,
          'eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWms9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--e9e436de22c082cd1142a42ea29daa7eaf5df97d',
        );
        expect(responseData.uploadUrl, '$serverUrl/file-upload');
        expect(responseData.headers['Content-MD5'], 'BdXPe8f6bkr3J4PcnYgnxw==');
      }).whenComplete(stopServer);
    });

    test('throws HttpStatusException on server error', () {
      return startServer()
          .then((_) async {
            var uploader = Uploader('$serverUrl/error');
            await expectLater(
              uploader.directUpload(directUploadRequestSample),
              throwsA(TypeMatcher<HttpStatusException>()),
            );
          })
          .catchError((_) {})
          .whenComplete(stopServer);
    });
  });

  group('directUpload()', () {
    test('sends large file in chunks and notifies progress', () async {
      var file = File('test/large.jpg');
      int fileSize = await file.length();
      List<double> progressPercents = [];
      return startServer().then((_) {
        var uploader = Uploader('$serverUrl/direct-upload');
        return uploader.fileUpload(
          fileContents: file.openRead(),
          byteSize: fileSize,
          directUploadResponse: correctDirectUploadResponse,
          onProgress: (percent) => progressPercents.add(percent),
        );
      }).then((_) {
        expect(progressPercents.length > 10, isTrue);
        expect(progressPercents.last, 100.0);
      }).whenComplete(stopServer);
    });

    test('sends small file altogether and notifies progress once', () async {
      var file = File('test/small.pdf');
      int fileSize = await file.length();
      List<double> progressPercents = [];
      return startServer().then((_) {
        var uploader = Uploader('$serverUrl/direct-upload');
        return uploader.fileUpload(
          fileContents: file.openRead(),
          byteSize: fileSize,
          directUploadResponse: correctDirectUploadResponse,
          onProgress: (percent) => progressPercents.add(percent),
        );
      }).then((_) {
        expect(progressPercents.length == 1, isTrue);
        expect(progressPercents.first, 100.0);
      }).whenComplete(stopServer);
    });

    test('throws HttpStatusException on server error', () async {
      var file = File('test/small.pdf');
      int fileSize = await file.length();
      return startServer()
          .then((_) async {
            var uploader = Uploader('$serverUrl/direct-upload');
            await expectLater(
              uploader.fileUpload(
                fileContents: file.openRead(),
                byteSize: fileSize,
                directUploadResponse: wrongDirectUploadResponse,
              ),
              throwsA(TypeMatcher<HttpStatusException>()),
            );
          })
          .catchError((_) {})
          .whenComplete(stopServer);
    });
  });
}

DirectUploadRequest get directUploadRequestSample => DirectUploadRequest(
      fileName: 'sample',
      contentType: 'text/plain',
      byteSize: 123456789,
      checksum: 'rL0Y20zC+Fzt72VPzMSk2A==',
    );

DirectUploadResponse get correctDirectUploadResponse => DirectUploadResponse(
      id: 153,
      key: 'zcYSmqkyKqJH6Mez7WJAcuJ8',
      signedId: 'eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWms9IiwiZXhwIjpudWxsL',
      uploadUrl: '$serverUrl/file-upload',
      headers: {'Content-MD5': 'BdXPe8f6bkr3J4PcnYgnxw=='},
    );

DirectUploadResponse get wrongDirectUploadResponse => DirectUploadResponse(
      id: 153,
      key: 'zcYSmqkyKqJH6Mez7WJAcuJ8',
      signedId: 'eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWms9IiwiZXhwIjpudWxsL',
      uploadUrl: '$serverUrl/error',
      headers: {'Content-MD5': 'BdXPe8f6bkr3J4PcnYgnxw=='},
    );


test/file_checksum_test.dart
import 'dart:io';

import 'package:active_storage/src/file_checksum.dart';
import 'package:test/test.dart';

void main() {
  test('returns md5 encoded as base64 string', () async {
    File file = File('test/sample.txt');
    String checksum = await FileChecksum.getMd5AsBase64(file: file);
    expect(checksum, 'rL0Y20zC+Fzt72VPzMSk2A==');
  });
}


lib/active_storage.dart
library active_storage;

import 'dart:io';

import 'src/direct_upload.dart';
import 'src/file_checksum.dart';
import 'src/uploader.dart';

export 'src/direct_upload.dart';
export 'src/exceptions.dart';

class ActiveStorage {
  final Uploader uploader;

  ActiveStorage({required String directUploadURL})
      : uploader = Uploader(directUploadURL);

  Future<DirectUploadResponse> upload({
    required String fileName,
    required String fileMimeType,
    required File file,
    ProgressCallback? onProgress,
  }) async {
    int fileSize = await file.length();
    String checksum = await FileChecksum.getMd5AsBase64(file: file);
    DirectUploadResponse response = await uploader.directUpload(
      DirectUploadRequest(
        fileName: fileName,
        contentType: fileMimeType,
        byteSize: fileSize,
        checksum: checksum,
      ),
    );
    await uploader.fileUpload(
      fileContents: file.openRead(),
      byteSize: fileSize,
      directUploadResponse: response,
      onProgress: onProgress,
    );
    return Future.value(response);
  }

  void addHeader(name, value) {
    uploader.addHeader(name, value);
  }
}


lib/src/exceptions.dart
class FileChecksumException implements Exception {
  final String? message;
  FileChecksumException(this.message);
}

class HttpStatusException implements Exception {
  String method;
  String? url;
  String? responseBody;
  String? requestBody;
  int? statusCode;

  HttpStatusException(
    this.method,
    this.url, {
    this.statusCode,
    this.requestBody,
    this.responseBody,
  });

  @override
  String toString() =>
      "HTTP error: $method $url\n\n$requestBody\n$responseBody";
}

class ClientError extends HttpStatusException {
  ClientError(
    String method,
    String? url,
    int statusCode,
    String? requestBody,
    String? responseBody,
  ) : super(
          method,
          url,
          statusCode: statusCode,
          requestBody: requestBody,
          responseBody: responseBody,
        );

  @override
  String toString() =>
      "Client request error: $method $url\n\n$requestBody\n$responseBody";
}

class ServerError extends HttpStatusException {
  ServerError(
    String method,
    String? url,
    int statusCode,
    String? requestBody,
    String? responseBody,
  ) : super(
          method,
          url,
          statusCode: statusCode,
          requestBody: requestBody,
          responseBody: responseBody,
        );

  @override
  String toString() =>
      "Server processing error: $method $url\n\n$requestBody\n$responseBody";
}

class NoNetworkError implements Exception {}

class NetworkError implements Exception {}

class InvalidDataReceived implements Exception {
  String message;
  InvalidDataReceived(this.message);
}


lib/src/direct_upload.dart
import 'dart:convert';

import 'exceptions.dart';

typedef ProgressCallback = void Function(double progressPercent);

class DirectUploadResponse {
  final int id;
  final String key;
  final String signedId;
  final String uploadUrl;
  final Map<String, String> headers;

  DirectUploadResponse({
    required this.id,
    required this.key,
    required this.signedId,
    required this.uploadUrl,
    required this.headers,
  });

  factory DirectUploadResponse.fromJson(String utf8Text) {
    try {
      var parsed = json.decode(utf8Text);
      return DirectUploadResponse(
        id: parsed['id'],
        key: parsed['key'],
        signedId: parsed['signed_id'],
        uploadUrl: parsed['direct_upload']['url'],
        headers: Map<String, String>.from(parsed['direct_upload']['headers']),
      );
    } on FormatException catch (e) {
      throw InvalidDataReceived(e.message);
    }
  }
}

class DirectUploadRequest {
  final String fileName;
  final String contentType;
  final int byteSize;
  final String checksum;

  DirectUploadRequest({
    required this.fileName,
    required this.contentType,
    required this.byteSize,
    required this.checksum,
  });

  String toJson() => json.encode({
        'blob': {
          'filename': fileName,
          'content_type': contentType,
          'byte_size': byteSize,
          'checksum': checksum,
        }
      });
}


lib/src/file_checksum.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:crypto/src/digest_sink.dart';

import 'exceptions.dart';

class FileChecksum {
  final Stream<List<int>> fileContents;
  late Digest md5Digest;

  static Future<String> getMd5AsBase64({required File file}) async {
    Stream<List<int>> _fileContents;
    try {
      _fileContents = file.openRead();
    } on UnsupportedError catch (e) {
      throw FileChecksumException(e.message);
    } on FileSystemException catch (e) {
      throw FileChecksumException(e.message);
    }
    FileChecksum checksum = FileChecksum(_fileContents);
    ;
    await checksum.compute();
    return checksum.base64;
  }

  FileChecksum(this.fileContents);

  String get base64 {
    try {
      return base64Encode(rawBytes);
    } on FormatException catch (e) {
      throw FileChecksumException(e.message);
    } on ArgumentError catch (e) {
      throw FileChecksumException(e.message);
    }
  }

  List<int> get rawBytes => md5Digest.bytes;

  Future<void> compute() async {
    var digestSink = DigestSink();
    try {
      ByteConversionSink md5Sink = md5.startChunkedConversion(digestSink);
      await for (List<int> chunk in fileContents) {
        md5Sink.add(chunk);
      }
      md5Sink.close();
      md5Digest = digestSink.value;
    } on UnsupportedError catch (e) {
      throw FileChecksumException(e.message);
    }
  }
}


lib/src/uploader.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import 'direct_upload.dart';
import 'exceptions.dart';

class Uploader {
  final String directUploadURL;
  var headers = Map<String, String>();

  Uploader(this.directUploadURL);

  Future<DirectUploadResponse> directUpload(
      DirectUploadRequest requestData) async {
    Map<String, String> requestHeaders = {
      'Content-Type': 'application/json; charset=utf-8',
      'Accept': 'application/json',
    };
    requestHeaders.addAll(headers);
    http.Response response =
        await _safelyRun<http.Response>(() async => await http.post(
              Uri.parse(directUploadURL),
              headers: requestHeaders,
              body: requestData.toJson(),
            ));
    String responseBody = _checkAndDecode(response);
    return DirectUploadResponse.fromJson(responseBody);
  }

  Future<void> fileUpload({
    required Stream<List<int>> fileContents,
    required int byteSize,
    required DirectUploadResponse directUploadResponse,
    ProgressCallback? onProgress,
  }) async {
    await _safelyRun<void>(() async {
      HttpClientRequest request =
          await HttpClient().putUrl(_getDirectUploadUri(directUploadResponse));
      directUploadResponse.headers.forEach((name, value) {
        request.headers.set(name, value);
      });
      request.headers.set('content-length', byteSize);
      int uploaded = 0;
      await request.addStream(fileContents.map((chunk) {
        uploaded += chunk.length;
        if (onProgress != null) onProgress(uploaded / byteSize * 100.0);
        return chunk;
      }));
      HttpClientResponse response = await request.close();
      if (response.statusCode < 200 || response.statusCode > 300) {
        String responseText = '';
        await for (String textChunk in utf8.decoder.bind(response)) {
          responseText += textChunk;
        }
        _throwHttpException(
          request.method,
          directUploadResponse.uploadUrl,
          statusCode: response.statusCode,
          responseBody: responseText,
        );
      }
    });
  }

  void addHeader(name, value) {
    headers[name] = value;
  }

  Uri _getDirectUploadUri(DirectUploadResponse directUploadResponse) {
    try {
      return Uri.parse(directUploadResponse.uploadUrl);
    } on FormatException catch (e) {
      throw InvalidDataReceived(e.message);
    }
  }

  String _checkAndDecode(http.Response response) {
    http.Request request = response.request as http.Request;
    String method = request.method;
    String url = request.url.toString();
    int code = response.statusCode;
    try {
      String responseBody = utf8.decode(response.bodyBytes);
      String requestBody = request.body;
      if (code < 200 || code > 300) {
        _throwHttpException(
          method,
          url,
          statusCode: code,
          requestBody: requestBody,
          responseBody: responseBody,
        );
      }
      return responseBody;
    } on FormatException catch (e) {
      throw InvalidDataReceived(e.message);
    }
  }

  void _throwHttpException(
    String method,
    String? url, {
    required int statusCode,
    String? requestBody,
    String? responseBody,
  }) {
    if (statusCode > 400 && statusCode < 500) {
      throw ClientError(method, url, statusCode, requestBody, responseBody);
    } else if (statusCode >= 500 && statusCode <= 599) {
      throw ServerError(method, url, statusCode, requestBody, responseBody);
    } else {
      throw HttpStatusException(
        method,
        url,
        statusCode: statusCode,
        requestBody: requestBody,
        responseBody: responseBody,
      );
    }
  }
}

Future<T> _safelyRun<T>(Future<T> method()) async {
  try {
    return await method();
  } on SocketException {
    throw NoNetworkError();
  } on http.ClientException {
    throw NetworkError();
  }
}
